src/netaddress.cpp

```diff
@@ -3,64 +3,88 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#include <cstdint>
 #include <netaddress.h>
 #include <hash.h>
 #include <util/strencodings.h>
 #include <util/asmap.h>
 #include <tinyformat.h>
 
-static const unsigned char pchIPv4[12] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0xff, 0xff };
-static const unsigned char pchOnionCat[] = {0xFD,0x87,0xD8,0x7E,0xEB,0x43};
+#include <algorithm>
+#include <cstdint>
+#include <iterator>
+#include <tuple>
+
+#ifdef WIN32
+#include <winsock.h> // htonl()
+#else
+#include <arpa/inet.h> // htonl()
+#endif
 
-// 0xFD + sha256("bitcoin")[0:5]
-static const unsigned char g_internal_prefix[] = { 0xFD, 0x6B, 0x88, 0xC0, 0x87, 0x24 };
+constexpr size_t CNetAddr::V1_SERIALIZATION_SIZE;
 
 /**
  * Construct an unspecified IPv6 network address (::/128).
  *
  * @note This address is considered invalid by CNetAddr::IsValid()
  */
-CNetAddr::CNetAddr()
-{
-    memset(ip, 0, sizeof(ip));
-}
+CNetAddr::CNetAddr() {}
 
 void CNetAddr::SetIP(const CNetAddr& ipIn)
 {
+    // Size check.
+    switch (ipIn.m_net) {
+    case NET_IPV4:
+        assert(ipIn.m_addr.size() == ADDR_IPV4_SIZE);
+        break;
+    case NET_IPV6:
+        assert(ipIn.m_addr.size() == ADDR_IPV6_SIZE);
+        break;
+    case NET_ONION:
+        assert(ipIn.m_addr.size() == ADDR_TORV2_SIZE);
+        break;
+    case NET_INTERNAL:
+        assert(ipIn.m_addr.size() == ADDR_INTERNAL_SIZE);
+        break;
+    case NET_UNROUTABLE:
+    case NET_MAX:
+        assert(false);
+    } // no default case, so the compiler can warn about missing cases
+
     m_net = ipIn.m_net;
-    memcpy(ip, ipIn.ip, sizeof(ip));
+    m_addr = ipIn.m_addr;
 }
 
-void CNetAddr::SetLegacyIPv6(const uint8_t ipv6[16])
+template <typename T, size_t PREFIX_LEN>
+inline bool HasPrefix(const T& obj, const uint8_t (&prefix)[PREFIX_LEN])
```

Comment:
I had to google this syntax :/ 
could we please use `const std::array<uint8_t, N>&` instead? :)

Reply:
or just make this generic over all containers

Reply:
> generic over all containers

Span !?

Reply:
I can't figure out how to use `std::array` or `Span` and keep the usage to just `HasPrefix(m_addr, {192, 168})` (which is sweet!), so I am keeping the reference to array.

Reply:
> I think that can be fixed by removing the T2 template parameter, and just hardcoding it to an array of uint8_t

> I think that can be fixed by removing the `T2` template parameter, and just hardcoding it to an array of `uint8_t` (but templated size).

It does not compile either because then gcc 4.8 complained it cannot deduce the length from the initialiser list, so I [switched](https://github.com/bitcoin/bitcoin/compare/f37cde9ad6764930c0fcfefb2139e6f07d6e4353..7c34237fba619667ed1d3981d3b564ad7c25077e) to `std::array` :-/

Reply:
This ended up with `const std::array<uint8_t, N>&` anyway, due to GCC 4.8. Will change it back to reference to an array once the support for GCC 4.8 is removed.